text/0082-dynamic-size-region.md

```diff
@@ -0,0 +1,189 @@
+on PD to split the hot spot in time and increase concurrency. In addition, we can also use unordered
+apply to make single region apply logs faster. Since unordered apply is a standalone feature, I’m not
+going into details here.
+
+For read hot spots, split should be triggered by PD, which can utilize the global statistics from all
+regions and nodes. For normal read requests, TiKV will need to split its range into smaller buckets
+according to statics to increase concurrency. When TiKV client wants to do a scan, it sends the RPC
+once, and TiKV will split the requests into smaller concurrent jobs.
+
+In the past design, follower read has also been discussed to offload works from leader. But client
+can’t predict the progress of a follower, so latency can also become unpredictable. It’s more
+controllable to let PD split hot spots and schedule to get balance.
+
+Now that a region can be in 10GiB, a full scan can make the response exceed the limit of gRPC,
+which is 4GiB. So instead of unary, we need to use server streaming RPC to return the response as
+stream. The stream can set a limit on memory or count to avoid too many pending responses
+exhausting memory.
+
+If follower read is enabled in client side, client can use buckets from PD to split requests and
+balance loads across leader and followers. Because buckets is not strictly mapping to a region,
+so client may maintain a sorted map about all available bucket keys, and update it in background
+periodically. When starting a query, client may query the map for available keys within region
+range and split request. When updating in the background, only removing overlap keys with the new
+region and keep the old, and possible stale, bucket keys. The point is that bucket keys don't need
+to be 100% accurate, a close match can enhance concurrency and load balance a lot. For fast lookup,
+a map from region id to bucket keys can also be maintained.
+
+After a region becomes larger, it needs to be warmed up before serving requests. There are two
+common cases:
+
+- In the very beginning, writes will only happen in a single region, the problem can be solved by
+    pre-split + scatter.
+- Because followers don’t serve read, so their data is usually not in cache. After transferring
+    the leader, serving a range scan can trigger at most 10GiB IO. To warm it up, we can extend
+    the existing pre transfer leader mechanism. Leader will send its bucket statics to follower
+    in pre transfer leader, follower can do a pre scan for hot buckets before accepting
+    transferring leader.
+
+### Flow report
+
+Buckets statistics are reported in a standalone stream. They will be reported more frequently
+than original region heartbeats, which is 60s. The new stream is expected to be reported every
+10 seconds when have changes.
+
+PD will maintain top N hot spot buckets and split all of those buckets that exceed P of
```

Comment:
I think 
- The Bucket needs to have a unique identifier. and it should like region split, use right derive. otherwise PD cannot know it is a stable hot bucket or it is a traffic peak.
- In most cases, it can be split from the boundary key of the bucket. but still need make load-base split key strategy or other strategy by PD can be accept.

Reply:
> ... it should like region split, use right derive

This can be very complicated. How about handling load trace on PD side? For example, if the buckets are [a, d), [d, f) in the past, and the new version is [a, c), [c, e), [e, f). Then history H[a, d) should be inherited by both [a, c) and [c, e), history H[d, f) is inherited by both [c, e) and [e, f). The overlapped histories should be sum up. And the inherited history statistic can be optionally multiply by a factor like 0.8 to make process smoothly.

> ...make load-base split key strategy or other strategy by PD...

It's up to PD to split by any keys. It's recommended to choose bucket boundary.

Reply:
I think this method also complicated. Do you mean that the bucket boundary may be different every time? [a, d)[d ,f) change to [a, c), [c, e), [e, f); d is gone.

Reply:
It can change, but it only changes when there are many updates.